### ISSUE TITLE
libtool is not filtered with Fujitsu compilers are used

### ISSUE THREAD

**Issue by ggouaillardet** [LLNL/spack#34897]
### Steps to reproduce

$ spack install --verbose testfilter %fj
[+] /vol0004/apps/oss/spack-v0.19/opt/spack/linux-rhel8-a64fx/fj-4.8.1/gnuconfig-2021-08-14-4m7rbpbv5rxc3rmornr7kh5556c2ajan
==> Installing testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp
==> No binary for testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp found: installing from source
==> Fetching https://github.com/ggouaillardet/testfilter/archive/refs/tags/v0.3.tar.gz
==> No patches needed for testfilter
==> testfilter: Executing phase: 'autoreconf'
==> testfilter: Executing phase: 'configure'
==> [2023-01-12-13:17:07.518990] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/configure [replacing "^(\s*if test x-L = )("\$p" \|\|\s*)$"]
==> [2023-01-12-13:17:07.544766] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/configure [replacing "^(\s*test x-R = )("\$p")(; then\s*)$"]
==> [2023-01-12-13:17:07.567225] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/configure [replacing "^(\s*test \$p = "-R")(; then\s*)$"]
==> [2023-01-12-13:17:07.592455] '/tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/configure' '--prefix=/vol0004/apps/oss/spack-v0.19/opt/spack/linux-rhel8-a64fx/fj-4.8.1/testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp'
==> [2023-01-12-13:17:07.715538] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "^wl=""$"]
==> [2023-01-12-13:17:07.736539] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "^pic_flag=""$"]
==> [2023-01-12-13:17:07.758015] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "^pic_flag=""$"]
==> [2023-01-12-13:17:07.780955] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "^pic_flag=""$"]
==> [2023-01-12-13:17:07.804096] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "^pic_flag=""$"]
==> [2023-01-12-13:17:07.827004] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "\-nostdlib"]
==> [2023-01-12-13:17:07.850862] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/fjhpctag\.o"]
==> [2023-01-12-13:17:07.874677] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/fjcrt0\.o"]
==> [2023-01-12-13:17:07.896767] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/fjlang08\.o"]
==> [2023-01-12-13:17:07.941023] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/fjomp\.o"]
==> [2023-01-12-13:17:07.965995] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/crti\.o"]
==> [2023-01-12-13:17:07.990996] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/crtbeginS\.o"]
==> [2023-01-12-13:17:08.018856] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/crtendS\.o"]
==> testfilter: Executing phase: 'build'
==> [2023-01-12-13:17:08.091422] 'make' '-j16' 'V=1'
if ! grep -qv crti.o libtool; then echo libtool was not filtered; exit 1; fi
libtool was not filtered
make: *** [Makefile:2: all] Error 1
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

1 error found in build log:
     18    ==> [2023-01-12-13:17:07.990996] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/crtbeginS\.o"]
     19    ==> [2023-01-12-13:17:08.018856] FILTER FILE: /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-src/libtool [replacing "/\\S\*/crtendS\.o"]
     20    ==> testfilter: Executing phase: 'build'
     21    ==> [2023-01-12-13:17:08.091422] 'make' '-j16' 'V=1'
     22    if ! grep -qv crti.o libtool; then echo libtool was not filtered; exit 1; fi
     23    libtool was not filtered
  >> 24    make: *** [Makefile:2: all] Error 1

See build log for details:
  /tmp/r00018/spack-stage/spack-stage-testfilter-0.3-auug2b6oxhttnilfjv3ylqp6h7vdhkyp/spack-build-out.txt

### Information on your system

$ spack debug report
* **Spack:** 0.19.0 (3a08b0506d40f8ed34d3bb3d5d24cf14b41ee07e)
* **Python:** 3.10.8
* **Platform:** linux-rhel8-a64fx
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by ggouaillardet** [comment]
The following `testfilter` package can be evidenced in order to reproduce the issue:

```
import os
import os.path

import llnl.util.lang

from spack import *


class Testfilter(AutotoolsPackage):
    """A dumb recipe that test FILTER"""

    url = "https://github.com/ggouaillardet/testfilter/archive/refs/tags/v0.1.tar.gz"

    version('0.3', sha256='95096a35d9b40b517be29078267846d8357ab37a65653e3266929d8dd925343c')
    version('0.2', sha256='e9b274b89e548d26fb850eff907d342ada70a351acbff94f53bd423d0bf1944c')
    version('0.1', sha256='83142d2e2882febc98b6a3f309b1b49e42901a3264c1ce6655e17322a093c866')
```

Fujitsu compilers must be requested in order to evidence the issue, but they are not used to compile anything, so a dummy `fj` environment can be used.

A workaround is the patch below:
```diff
diff --git a/lib/spack/spack/build_systems/autotools.py b/lib/spack/spack/build_systems/autotools.py
index 76e6bb27ce..c66591cd7b 100644
--- a/lib/spack/spack/build_systems/autotools.py
+++ b/lib/spack/spack/build_systems/autotools.py
@@ -424,7 +424,7 @@ def _do_patch_libtool(self):
 
         # Other compiler-specific patches:
         if self.pkg.compiler.name == "fj":
-            x.filter(regex="-nostdlib", repl="", string=True)
+            x.filter(regex="-nostdlib", repl="")
             rehead = r"/\S*/"
             for o in [
                 "fjhpctag.o",
@@ -435,7 +435,7 @@ def _do_patch_libtool(self):
                 "crtbeginS.o",
                 "crtendS.o",
             ]:
-                x.filter(regex=(rehead + o), repl="", string=True)
+                x.filter(regex=(rehead + o), repl="")
         elif self.pkg.compiler.name == "dpcpp":
             # Hack to filter out spurious predep_objects when building with Intel dpcpp
             # (see https://github.com/spack/spack/issues/32863):
```

The changes were introduced in #4bc8f663881.

@skosukhin @alalazo @haampie
What is the rationale for `x.filter(... string=True)`?
If this is not needed here, it might also not be needed for other compilers I did not test.

**Comment by skosukhin** [comment]
@ggouaillardet thank you for the report!

1. The `string` argument is documented [here](https://github.com/spack/spack/blob/03fc502da56bd1cdf17acb4da6e2c2806fa3bf87/lib/spack/llnl/util/filesystem.py#L243).
2. `string=True` [here](https://github.com/spack/spack/blob/03fc502da56bd1cdf17acb4da6e2c2806fa3bf87/lib/spack/spack/build_systems/autotools.py#L438) is a bug. My bad, sorry. The argument should be removed because `rehead` is actually a regular expression. Would you create a pull request fixing this?
3. `string=True` [here](https://github.com/spack/spack/blob/03fc502da56bd1cdf17acb4da6e2c2806fa3bf87/lib/spack/spack/build_systems/autotools.py#L427) should stay, in my opinion, because `-nostdlib` is not a regular expression but a plain string.

**Comment by ggouaillardet** [comment]
Thanks @skosukhin for the quick reply!

I filed #34916 (against the `develop` branch) in order to fix this

Once merged, I will open a PR vs the `releases/v0.19` branch

**Comment by alalazo** [comment]
Added #34916 to the list in https://github.com/spack/spack/pull/33966 , so closing this issue
